### PR TITLE
Fix the bug that distribution goals may move more than necessary replicas in self-healing scenario.

### PR DESCRIPTION
**Pull request by kidkun**
Fix issue #809

### CODE REVIEW

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/OptimizationVerifier.java [hunk 1]
```diff
@@ -122,18 +127,34 @@ static boolean executeGoalsFor(BalancingConstraint constraint,
                                                     new SystemTime(),
                                                     new MetricRegistry(),
                                                     EasyMock.mock(Executor.class));
-    GoalOptimizer.OptimizerResult optimizerResult = goalOptimizer.optimizations(clusterModel,
-                                                                                goalByPriority,
-                                                                                new OperationProgress());
+    // First optimize with all hard goals if there is any.
+    List<Goal> hardGoals =  goalByPriority.stream().filter(Goal::isHardGoal).collect(Collectors.toList());
+    GoalOptimizer.OptimizerResult optimizerResult1 = hardGoals.isEmpty() ? null
```
Comment (efeg) [inline cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/OptimizationVerifier.java]: `optimizerResult1` -> `optimizationResultWithHardGoals`?

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/OptimizationVerifier.java [hunk 2]
```diff
@@ -122,18 +127,34 @@ static boolean executeGoalsFor(BalancingConstraint constraint,
...
+                                                                         : goalOptimizer.optimizations(clusterModel,
+                                                                                                       hardGoals,
+                                                                                                       new OperationProgress());
+
+    // If provided goal list contains any soft goals, then optimize with all goals.
+    // In this iteration, all the hard goals will not generate any new proposal since they are already satisfied in previous
+    // iteration (otherwise an exception has already been thrown). The optimize result of this iteration will only contains
+    // proposals generated by soft goals.
+    List<Goal> softGoals = goalByPriority.stream().filter(g -> !g.isHardGoal()).collect(Collectors.toList());
+    GoalOptimizer.OptimizerResult optimizerResult2 = softGoals.isEmpty() ? null
```
Comment (efeg) [inline cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/OptimizationVerifier.java]: `optimizerResult1` -> `optimizationResultWithSoftGoals`?

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/OptimizationVerifier.java [hunk 3]
```diff
@@ -122,18 +127,34 @@ static boolean executeGoalsFor(BalancingConstraint constraint,
...
+    List<Goal> hardGoals =  goalByPriority.stream().filter(Goal::isHardGoal).collect(Collectors.toList());
+    GoalOptimizer.OptimizerResult optimizerResult1 = hardGoals.isEmpty() ? null
+                                                                         : goalOptimizer.optimizations(clusterModel,
+                                                                                                       hardGoals,
+                                                                                                       new OperationProgress());
+
+    // If provided goal list contains any soft goals, then optimize with all goals.
+    // In this iteration, all the hard goals will not generate any new proposal since they are already satisfied in previous
+    // iteration (otherwise an exception has already been thrown). The optimize result of this iteration will only contains
+    // proposals generated by soft goals.
```
Comment (efeg) [inline cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/OptimizationVerifier.java]: The explanation is correct, but this change might introduce potential issues:
1. We no longer test the optimization process with a mix of soft and hard goals.
2. In particular, we no longer test the cases, where soft goals have higher priority compared to hard goals.
-- 
Does it make sense to:
(1) preserve the existing test cases with the original list of goals and
(2) introduce "selected" (i.e. targeted) test cases for 
    2.1. testing each soft goal in a cluster state such as the one exemplified in the corresponding ticket of this PR? -- i.e. a case that the goal needs to move non-immigrant replica(s) to achieve balance. But, for correctness, it should not move.
    2.2. testing each hard goal in a cluster state such that the goal needs to move non-immigrant replica(s) to achieve balance. And, for correctness, it indeed moves replicas.

Comment (kidkun) [reply cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/OptimizationVerifier.java]: Good point, I am adding a new switch to `OptimizationVerifier` to specify wether optimize with provided goals as a whole or optimize in a two-pass way(first optimize with all hard goals, then optimize with all soft goals).

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/OptimizationVerifier.java [hunk 4]
```diff
@@ -93,12 +95,38 @@ static boolean executeGoalsFor(BalancingConstraint constraint,
...
+  }
+
+  /**
+   * Execute given goals in the given cluster enforcing the given constraint. Return pass / fail status of a test.
+   * A test fails if:
+   * 1) Rebalance: During the optimization process, optimization of a goal leads to a worse cluster state (in terms of
+   * the requirements of the same goal) than the cluster state just before starting the optimization.
+   * 2) Self Healing: There are replicas on dead brokers after self healing or soft goal(if any) moves non-immigrant
+   * online replicas.
+   * 3) Adding a new broker causes the replicas to move among old brokers.
```
Comment (efeg) [inline cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/OptimizationVerifier.java]: Nit: Looks like the following JavaDoc is repeated in each `#executeGoalsFor`:
```
   * Execute given goals in the given cluster enforcing the given constraint. Return pass / fail status of a test.
   * A test fails if:
   * 1) Rebalance: During the optimization process, optimization of a goal leads to a worse cluster state (in terms of
   * the requirements of the same goal) than the cluster state just before starting the optimization.
   * 2) Self Healing: There are replicas on dead brokers after self healing or soft goal(if any) moves non-immigrant
   * online replicas.
   * 3) Adding a new broker causes the replicas to move among old brokers.
```
For conciseness, should we refer to this JavaDoc from the other two `#executeGoalsFor` for the excerpt above? -- i.e. we can say something like:
```
See {@link #executeGoalsFor(BalancingConstraint, ClusterModel, List, Collection, List, boolean)}.
```

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/DeterministicSelfHealingTest.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+                                                    TopicReplicaDistributionGoal.class.getName());
+
+    Properties props = KafkaCruiseControlUnitTestUtils.getKafkaCruiseControlProperties();
+    props.setProperty(KafkaCruiseControlConfig.MAX_REPLICAS_PER_BROKER_CONFIG, Long.toString(5L));
+    BalancingConstraint balancingConstraint = new BalancingConstraint(new KafkaCruiseControlConfig(props));
+    balancingConstraint.setCapacityThreshold(TestConstants.LOW_CAPACITY_THRESHOLD);
+    balancingConstraint.setResourceBalancePercentage(TestConstants.LOW_BALANCE_PERCENTAGE);
+    List<OptimizationVerifier.Verification> verifications = Arrays.asList(DEAD_BROKERS, REGRESSION);
+
+    // ----------##TEST 1: SINGLE GOAL.
```
Comment (efeg) [inline cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/DeterministicSelfHealingTest.java]: Don't we already have the same tests with a single goal in `RandomSelfHealingTest`? If they are the same, can we remove the test with single goal and keep only the `RACK AWARE GOAL + ONE OTHER GOAL` part?
Finally, can we rename this test class to a more descriptive name -- e.g. `DeterministicSelfHealingTest` -> `SelfHealingWithOptimizedGoal`?

**Pull request merged**